### ISSUE TITLE
Unhandled Exception: Bad state: Expected exactly one element, but got 2

### ISSUE THREAD

**Issue by javdhu** [simolus3/drift#1506]
I seem to get this bug on a table with a joint primary key. I use getSingleOrNull to get one element and it returns two. The two returned have different primary keys even though I explicitly query by one primary key.
This is the exact error
[ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: Bad state: Expected exactly one element, but got 2

**Comment by Rossdex** [comment]
I had a similar issue to this and resolved it by putting a limit on my queries where I know I only want 1. 

```
  Stream<D?> getBy(String key, dynamic val) {
    CustomExpression<bool> findByExpression = CustomExpression<bool>("$key = '$val'");
    int? userId = db.sharedPreferences.getInt('current_user_id');

    return (select(_table)
          ..where((tbl) => findByExpression)
          ..where((tbl) => tbl.currentUserId.equals(userId))
          ..orderBy([(tbl) => OrderingTerm(expression: tbl.id, mode: OrderingMode.desc)])
          ..limit(1))
        .watchSingleOrNull();
  }
```

**Comment by North101** [comment]
```
CREATE TABLE folder_entries (
    id TEXT NOT NULL,
    driveId TEXT NOT NULL,

    name TEXT NOT NULL,
    parentFolderId TEXT,
    path TEXT NOT NULL,

    dateCreated DATETIME NOT NULL DEFAULT (strftime('%s','now')),
    lastUpdated DATETIME NOT NULL DEFAULT (strftime('%s','now')),

    PRIMARY KEY (id, driveId)
) As FolderEntry;
```
The primary key is `id` and `driveId`

```
folderById:
    SELECT * FROM folder_entries
    WHERE driveId = :driveId AND id = :folderId;
```
You're where clause is filtering on `driveId` and `folderId`.

It's possible that theres duplicate entries in `folder_entries`.

**Comment by srix55** [comment]
UPDATE: it's got nothing to do with isolates. Rather, it's got something to do with nullable columns in a compound primary key.

I'm running into this issue too. Seems like two insertions are happening even if they violate the primary key constraint. It seems to be happening when insertions are made to the same database over different isolates simultaneously. Violating insertions in the same isolate doesn't cause duplicates. I am using worker_manager for managing isolates & hooking up drift for isolates using the pre flutter 3.7 way.

Let me build a reproducible code.

**Comment by srix55** [comment]
I realize that none of the databases (mysql & postgresql) allow nullability in any of the columns of a compound primary key. That is primary key is always non-nullable - be it a single column or multiple - all columns have to be non-nullable.
- Drift allows columns of a compound primary key to be nullable
- Drift doesn't guarantee primary key uniqueness when same records (with null in nullable columns) are inserted

Here's the reproducible code: https://github.com/srix55/drift_primary_key_issue

**Comment by simolus3** [comment]
With sqlite3, it's [possible to have nullable primary keys](https://sqlite.org/lang_createtable.html#the_primary_key). As the docs mention, that is more of a defect in sqlite3 than anything else, but that's what it is.
Two nulls being counted as distinct values in primary key is also the intended behavior. 

I wonder what the best course of action here is - adding a warning probably?

**Comment by srix55** [comment]
True... it's an SQLite bug. Many others are facing the same - 
https://github.com/prisma/prisma/issues/3130
https://stackoverflow.com/questions/64753105/why-can-i-add-null-value-to-primary-key-in-sqlite

Drift should not be addressing this, imo. Let's make sure not to declare nullable primary keys.

**Comment by simolus3** [comment]
Yeah. I'll consider making `STRICT` the default in a future drift version, which would fix this. 

The original issue description had all involved columns declared as non-null though, so it's possible that something else is/was going wrong here as well.
